File: python/linked_list/q19.py
"""
Given a linked list, remove the nth node from the end of list and return its head.

For example,

   Given linked list: 1->2->3->4->5, and n = 2.

   After removing the second node from the end, the linked list becomes 1->2->3->5.

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
"""
def removeNthFromEnd(self, head, n):
    """
    :type head: ListNode
    :type n: int
    :rtype: ListNode
    """
    dummy = ListNode(0)
    dummy.next = head

    f = dummy
    s = dummy

    for i in range(0,n+1):
        f = f.next
    while(f != None):
        f = f.next
        s = s.next

    temp = s.next
    s.next = s.next.next
    del temp
    return dummy.next

File: c++/linked_list/q19.cpp
ListNode* removeNthFromEnd(ListNode* head, int n) {
    ListNode* dummy = new ListNode(0);
    dummy->next = head;
    ListNode* f = dummy; ListNode* s = dummy;

    for(int i = 0; i <= n; i++)
        f = f->next;

    while(f != nullptr){
        f = f->next;
        s = s->next;
    }

    ListNode* temp = s->next;
    s->next = s->next->next;
    delete temp;
    return dummy->next;
}
